### PR TITLE
Rename `gcs.endpoint` to `gcs.service.host`

### PR DESCRIPTION
**Pull request by Fokko**
To make it in line with Java:

https://github.com/apache/iceberg/blob/6ee6d1327d3811dbd5795c4e87efdc41b7a58eaa/gcp/src/main/java/org/apache/iceberg/gcp/GCPProperties.java#L32

### CODE REVIEW

File: pyiceberg/io/fsspec.py [hunk 1]
```diff
@@ -158,6 +160,14 @@ def _gs(properties: Properties) -> AbstractFileSystem:
     # https://gcsfs.readthedocs.io/en/latest/api.html#gcsfs.core.GCSFileSystem
     from gcsfs import GCSFileSystem
 
+    if (endpoint := properties.get(GCS_ENDPOINT)) and GCS_SERVICE_HOST not in properties:
+        deprecated(
```
Comment (ndrluis) [inline pyiceberg/io/fsspec.py:164]: Now we have the deprecation_message function, which could be used here.

File: pyiceberg/io/fsspec.py [hunk 2]
```diff
@@ -158,6 +160,14 @@ def _gs(properties: Properties) -> AbstractFileSystem:
...
+            deprecated_in="0.8.0",
+            removed_in="0.9.0",
+            help_message=f"The property {GCS_ENDPOINT} is deprecated, please use {GCS_SERVICE_HOST} instead",
+        )(lambda: None)()
```
Comment (ndrluis) [inline pyiceberg/io/fsspec.py:168]:
```suggestion
        deprecation_message(
            deprecated_in="0.8.0",
            removed_in="0.9.0",
            help_message=f"The property {GCS_ENDPOINT} is deprecated, please use {GCS_SERVICE_HOST} instead",
        )
```

File: pyiceberg/io/pyarrow.py [hunk 3]
```diff
@@ -388,7 +389,14 @@ def _initialize_fs(self, scheme: str, netloc: Optional[str] = None) -> FileSyste
                 gcs_kwargs["credential_token_expiration"] = millis_to_datetime(int(expiration))
             if bucket_location := self.properties.get(GCS_DEFAULT_LOCATION):
                 gcs_kwargs["default_bucket_location"] = bucket_location
-            if endpoint := self.properties.get(GCS_ENDPOINT):
+            if (endpoint := self.properties.get(GCS_ENDPOINT)) and GCS_SERVICE_HOST not in self.properties:
+                deprecated(
+                    deprecated_in="0.8.0",
+                    removed_in="0.9.0",
+                    help_message=f"The property {GCS_ENDPOINT} is deprecated, please use {GCS_SERVICE_HOST} instead",
+                )(lambda: None)()
```
Comment (ndrluis) [inline pyiceberg/io/pyarrow.py:397]:
```suggestion
                deprecation_message(
                    deprecated_in="0.8.0",
                    removed_in="0.9.0",
                    help_message=f"The property {GCS_ENDPOINT} is deprecated, please use {GCS_SERVICE_HOST} instead",
                )
```

File: pyiceberg/io/fsspec.py [hunk 4]
```diff
@@ -76,6 +77,7 @@
     OutputStream,
 )
 from pyiceberg.typedef import Properties
+from pyiceberg.utils.deprecated import deprecated
```
Comment (ndrluis) [inline pyiceberg/io/fsspec.py:80]:
```suggestion
from pyiceberg.utils.deprecated import deprecation_message
```

File: pyiceberg/io/fsspec.py (hunk 1 above)
Comment (Fokko) [reply pyiceberg/io/fsspec.py]: Yes, that's much nicer, thanks for adding that @ndrluis

**Comment by ndrluis** [comment]
@Fokko, I can solve the missing import. Do you mind?

File: pyiceberg/io/fsspec.py [hunk 5]
```diff
@@ -179,7 +188,7 @@ def _gs(properties: Properties) -> AbstractFileSystem:
         cache_timeout=properties.get(GCS_CACHE_TIMEOUT),
         requester_pays=property_as_bool(properties, GCS_REQUESTER_PAYS, False),
         session_kwargs=json.loads(properties.get(GCS_SESSION_KWARGS, "{}")),
-        endpoint_url=properties.get(GCS_ENDPOINT),
+        endpoint_url=properties.get(GCS_SERVICE_HOST),
```
Comment (kevinjqliu) [inline pyiceberg/io/fsspec.py:191]:
```suggestion
        endpoint_url=get_first_property_value(properties, GCS_SERVICE_HOST, GCS_ENDPOINT),
```

File: pyiceberg/io/fsspec.py [hunk 6]
```diff
@@ -171,6 +172,14 @@ def _gs(properties: Properties) -> AbstractFileSystem:
...
     from gcsfs import GCSFileSystem
 
+    if (endpoint := properties.get(GCS_ENDPOINT)) and GCS_SERVICE_HOST not in properties:
+        deprecation_message(
+            deprecated_in="0.8.0",
+            removed_in="0.9.0",
+            help_message=f"The property {GCS_ENDPOINT} is deprecated, please use {GCS_SERVICE_HOST} instead",
+        )
+        properties[GCS_SERVICE_HOST] = endpoint
+
```
Comment (kevinjqliu) [inline pyiceberg/io/fsspec.py:182]:
```suggestion
    if properties.get(GCS_ENDPOINT):
        deprecation_message(
            deprecated_in="0.8.0",
            removed_in="0.9.0",
            help_message=f"The property {GCS_ENDPOINT} is deprecated, please use {GCS_SERVICE_HOST} instead",
        )
```

**Review by kevinjqliu — APPROVED** [review]
LGTM! Follows the java convention https://iceberg.apache.org/javadoc/0.14.1/constant-values.html

wdyt about using `get_first_property_value`

**Comment by Fokko** [comment]
Thanks @kevinjqliu, using `get_first_property_value` is much cleaner

File: pyiceberg/io/pyarrow.py [hunk 7]
```diff
@@ -390,7 +391,14 @@ def _initialize_fs(self, scheme: str, netloc: Optional[str] = None) -> FileSyste
...
                 gcs_kwargs["default_bucket_location"] = bucket_location
-            if endpoint := self.properties.get(GCS_ENDPOINT):
+            if (endpoint := self.properties.get(GCS_ENDPOINT)) and GCS_SERVICE_HOST not in self.properties:
+                deprecation_message(
+                    deprecated_in="0.8.0",
+                    removed_in="0.9.0",
+                    help_message=f"The property {GCS_ENDPOINT} is deprecated, please use {GCS_SERVICE_HOST} instead",
+                )
+                self.properties[GCS_SERVICE_HOST] = endpoint
+            if endpoint := self.properties.get(GCS_SERVICE_HOST):
```
Comment (kevinjqliu) [inline pyiceberg/io/pyarrow.py:401]:
```suggestion
            if endpoint := get_first_property_value(self.properties, GCS_SERVICE_HOST, GCS_ENDPOINT):
                if self.properties.get(GCS_ENDPOINT):
                    deprecation_message(
                        deprecated_in="0.8.0",
                        removed_in="0.9.0",
                        help_message=f"The property {GCS_ENDPOINT} is deprecated, please use {GCS_SERVICE_HOST} instead",
                    )
```
wdyt?

**Review by kevinjqliu — APPROVED** [review]

**Pull request merged**